torch/utils/data/_utils/worker.py

```diff
@@ -134,8 +134,13 @@ def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
         signal_handling._set_worker_signal_handlers()
 
         torch.set_num_threads(1)
+        seed = base_seed + worker_id
         random.seed(seed)
         torch.manual_seed(seed)
+        if HAS_NUMPY:
+            import numpy as np
+            ss = np.random.SeedSequence([worker_id, base_seed])
```

Comment:
@ejguan you're dropping compatibility with `numpy < 1.17.0` here. That probably wasn't intended. This is _just_ okay for conda-forge, but in the `builder` repo lower versions of numpy specified.

Since all you need is an array of four `uint32` numbers, I recommend checking `np.__version__` and if it's lower than `'1.17'` provide a fallback.

Reply:
Just be careful with the fallback. Don't reuse the same seed for the stdlib `random` and `np.random`, for example. [Bad things happen.](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659)

I wouldn't say that "all you need is _an_ array of `uint32` numbers". Safely deriving parallel streams is a tricky business. The PRNGs in play here (MT19937 and MTGP) just don't have the properties that we would want them to in order to use simple seed derivation schemes. What particular fallback were you thinking of?

Reply:
@rgommers Oh, thank you for bringing it up. Will have a patch to fix it.

Reply:
@rkern Thanks for providing insights on seeding for RNG. Just wondering if it falls back to an array of four `uint32` number, which way do you suggest to generate these four numbers?

Reply:
Honestly, I suggest vendorizing `SeedSequence` (or its [C++ progenitor](https://gist.github.com/imneme/540829265469e673d045)).

Reply:
I am thinking, at least for now, to utilize the 64bit int from torch seed for numpy to have a better random performance

Reply:
Ah, that was the [Bad things happen](https://github.com/PyTorchLightning/pytorch-lightning/pull/6960#issuecomment-818393659) option that I warn against. :-)

Both the stdlib `random` and `np.random` (pre 1.17) both use the same MT19937 PRNG algorithm, and their seeding algorithms are the same (largely). You do not want them both to start with the same state because both might get used in the worker (I see a lot of code that mixes `random` and `np.random` in ML workflows). You want both of those PRNG streams to be decorrelated from each other. The integer hashes underlying `SeedSequence` give you a good way to deterministically derive high-quality sets of seeds derived from sets that are convenient (e.g. `base_seed+i`) but merely distinct. Distinctness isn't _quite_ enough for MT19937 the same way it is for Philox. So, for example, I also wouldn't try to use 

```python
random.seed(torch_seed)
np.random.seed([torch_seed % (1<<32), torch_seed >> 32, SOME_CONSTANT_TO_MAKE_THEM_DISTINCT])
```

The MT19937 seeding algorithm is _okay_, but it wasn't really designed to be stress-tested in this way to generate parallel streams. `SeedSequence` was.

Reply:
Understood, same seed leads to same result with same algorithm. And, you are right. I think the ultimate goal or solution is to have `SeedSequence` implemented in PyTorch. But, before that, I can only provide a solution by making sure the distinctness, but with better seeding algorithm and randomness for `numpy.random` and `random` module. Does that make sense?

Reply:
I'm not sure what's stopping you from vendorizing `SeedSequence` now.

But if you need a quick stopgap, you can pass the 64-bit seed through [SplitMix64](https://xoshiro.di.unimi.it/splitmix64.c) or some other similar [64-bit integer hash](https://nullprogram.com/blog/2018/07/31/) (there are a lot that use this pattern, but do make sure to use the initial addition that SplitMix64 does to prevent 0 from mapping to 0).

```python
M64 = 0xFFFFFFFF_FFFFFFFF
def splitmix64(u64):
    z = (u64 + 0x9e3779b97f4a7c15) & M64
    z = (z ^ (z >> 30)) & M64
    z = (z * 0xbf58476d1ce4e5b9) & M64
    z = (z ^ (z >> 27)) & M64
    z = (z * 0x94d049bb133111eb) & M64
    z = (z ^ (z >> 31)) & M64
    return z

torch_seed = torch.initial_seed()
# We can reuse the PyTorch's seed once because it uses Philox and random uses MT19937.
random.seed(torch_seed)
# But we can't reuse it a second time for np.random since it also uses MT19937.
# Permute it deterministically to ensure that it is "far" from torch_seed.
numpy_seed = splitmix64(torch_seed)
np.random.seed([numpy_seed % (1<<32), numpy_seed >> 32])
```

Reply:
We will utilize SeedSequence for NumPy >= 1.17.0, but it's not available for prior NumPy releases

Reply:
Right, that's why I recommended vendorizing it.

Reply:
I have no objection to have SeedSequence. But, I prefer fixing it as it's a bug.

Reply:
Vendorizing `SeedSequence` (i.e. copying the [`SeedSequence` implementation](https://github.com/numpy/numpy/blob/main/numpy/random/bit_generator.pyx#L252-L476) into PyTorch or reimplementing the algorithm in whatever way is convenient for you) would fix the bug. Importantly, it would fix it the same way for all versions of numpy and not have two different behaviors depending on which version of numpy the user had installed.

Reply:
I was thinking to derive the numbers via a simple scheme from the pytorch PRNG indeed. However, I agree that vendoring `SeedSequence` is better. Should be the C++ progenitor you linked to above, because PyTorch doesn't have a build dependency on Cython, and that would be annoying to add. Or just translate the `SeedSequence` implementation to pure Python; performance for a one-off call is going to be fine compared to loading lots of data.

Reply:
Yeah, the only wrinkle to implementing `SeedSequence` in Python is that you have to be explicit about enforcing the `uint32` clock arithmetic.

Reply:
One PR #56797 is submitted with a hard-code state generation for NumPy. And an issue #56805 is also submitted for feature request of `SeedSequence`. Feel free to review and comment on them. Thanks a lot!